torch/_inductor/simple_fsdp/greedy_bucket.py

```diff
@@ -0,0 +1,379 @@
+import math
+from enum import IntEnum
+from collections import defaultdict
+from typing import Dict, List, Optional, Set, Tuple, Union
+
+import torch
+import torch.distributed as dist
+from .. import dependencies, ir, scheduler, config
+
+from .utils import NodeType, compute_node_users, get_node_type
+from .bucket import merge_allgather, merge_ag_wait, merge_reducescatter, merge_rs_wait
+
+
+def bucket_forward(
+    sched: "scheduler.Scheduler",
+    snodes: List["scheduler.BaseSchedulerNode"],
+    run_time_dict: Dict[str, List[Union[str, float, float]]],
+) -> List["scheduler.BaseSchedulerNode"]:
+    """
+    Greedy bucket ALL_GATHER and AG_WAIT
+    """
+    node_dep_dict = {}
+
+    inverse_users, node_users = compute_node_users(snodes)
+    comp = []
+    skip_nodes = []
+    all_gather = None
+    total_dep_list = []
+    for node in snodes:
+        if node.get_name() in run_time_dict:
+            _, run_time, memory = run_time_dict[node.get_name()]
+        else:
+            run_time, memory = 0, 0
+
+        if get_node_type(node) == NodeType.ALL_GATHER:
+            # A ag_dict consists of 6 parts:
+            # [Node]: (1) AG_DEP: The node ALL_GATHER depends on for read-in; (2) AG_WAIT: ALL_GATHER's wait nodes; (3) COMPUTE: COMPUTE nodes ALL_GATHER fetches.
+            # [Estimation Num.]: (1) AG_TIME: The estimated ALL_GATHER time; (2) COMPUTE_TIME: The estimated COMPUTE time; (3) MEMORY: The estimated meory for computation.
+            ag_dict = forward_ag_dict()
+            ag_wait = list(node_users[node])
+            assert [get_node_type(i) == NodeType.AG_WAIT for i in ag_wait] 
```

Comment:
why there can be multiple `wait` for the AG?
aslo I believe you mean something like `assert all([...])`